[Docs] pdn documentation (#3710)
The-OpenROAD-Project/OpenROAD: src/pdn/README.md

> # PDN Analysis

Reviewer: It's a power grid generator, not an analysis tool

> The power distribution network (PDN) analysis module in OpenROAD (`pdn`) 

Reviewer: generator

| ----- | ----- |
| `-macro` | Defines the type of grid being added as a macro. |
| `-name` | Defines a name to use when referring to this grid definition. |
| `-grid_over_pg_pins`, `-grid_over_boundary` | Place the power grid over the power ground pins of the macro. (Default True), or Place the power grid over the entire macro.  |
| `-voltage_domain` | Defines the name of the voltage domain for this grid. (Default: Last domain created). |
| `-orient` | For a macro, defines a set of valid orientations. LEF orientations (N, FN, S, FS, E, FE, W and FW) can be used as well as standard geometry orientations (R0, R90, R180, R270, MX, MY, MXR90 and MYR90). Macros with one of the valid orientations will use this grid specification. |

> | `-instances` | For a macro, defines a set of valid instances. Macros with a matching instance name will use this grid specification. |
> | `-cells` | For a macro, defines a set of valid cells. Macros which are instances of one of these cells will use this grid specification. |
> | `-default` | For a macro, specifies this is a default grid that can be overwritten. |

Reviewer 2: ```suggestion
| `-instances` | For a macro, defines a set of valid instances. Macros with a matching instance name will use this grid specification. |
| `-cells` | For a macro, defines a set of valid cells. Macros, which are instances of one of these cells, will use this grid specification. |
| `-default` | For a macro, specifies this is a default grid that can be overwritten. |
```

> | `-halo` | Specifies the default minimum separation of selected macros from other cells in the design. This is only used if the macro does not define halo values in the LEF description. If 1 value is specified it will be used on all 4 sides, if two values are specified, the first will be applied to left/right sides and the second will be applied to top/bottom sides, if 4 values are specified, then they are applied to left, bottom, right and top sides respectively (Default: 0). |
> | `-pins` | Defines a list of layers which where the power straps will be promoted to block pins. |

Reviewer 2: ```suggestion
| `-halo` | Specifies the design's default minimum separation of selected macros from other cells. This is only used if the macro does not define halo values in the LEF description. If one value is specified, it will be used on all four sides; if two values are specified, the first will be applied to left/right sides, and the second will be applied to top/bottom sides; if four values are specified, then they are applied to left, bottom, right and top sides respectively (Default: 0). |
| `-pins` | Defines a list of layers where the power straps will be promoted to block pins. |
```

| `-starts_with` | Specifies whether the first strap placed will be POWER or GROUND (Default: GROUND). |

> | `-obstructions` | Specify the layers to add routing blockages, in order to avoid DRC violations. |
> | `-power_switch_cell` | Defines the name of the coarse grain power switch cell to be used wherever the stdcell rail connects to the rest of the power grid. The mesh layers are associated with the unswitched power net of the voltage domain, whereas the stdcell rail is associated with the switched power net of the voltage domain. The placement of a power switch cell connects the unswitched power mesh to the switched power rail through a power switch defined by the `define_power_switch_cell` command. |
> | `-power_control` | Defines the name of the power control signal used to control the switching of the inserted power switches. |
> | `-power_control_network` | Defines the structure of the power control signal network. Choose from STAR, or DAISY. If STAR is specified, then the network is wired as a high-fanout net with the power control signal driving the power control pin on every power switch. If DAISY is specified then the power switches are connected in a daisy-chain configuration - note, this requires that the power swich defined by the `define_power_switch_cell`  command defines an acknowledge pin for the switch. |

Reviewer 2: ```suggestion
| `-obstructions` | Specify the layers to add routing blockages in order to avoid DRC violations. |
| `-power_switch_cell` | Defines the name of the coarse grain power switch cell to be used wherever the stdcell rail connects to the rest of the power grid. The mesh layers are associated with the unswitched power net of the voltage domain, whereas the stdcell rail is associated with the switched power net of the voltage domain. The placement of a power switch cell connects the unswitched power mesh to the switched power rail through a power switch defined by the `define_power_switch_cell` command. |
| `-power_control` | Defines the name of the power control signal used to control the switching of the inserted power switches. |
| `-power_control_network` | Defines the structure of the power control signal network. Choose from STAR or DAISY. If STAR is specified, then the network is wired as a high-fanout net with the power control signal driving the power control pin on every power switch. If DAISY is specified, then the power switches are connected in a daisy-chain configuration - note, this requires that the power switch defined by the `define_power_switch_cell`  command defines an acknowledge pin for the switch. |
```

Example usage:

```tcl
define_pdn_grid -macro -name ram          -orient {R0 R180 MX MY} -grid_over_pg_pins  -starts_with POWER -pin_direction vertical
define_pdn_grid -macro -name rotated_rams -orient {E FE W FW}     -grid_over_boundary -starts_with POWER -pin_direction horizontal
```

### Define Power Grid for an existing routing solution

```tcl
define_pdn_grid 
    -existing
    [-name name]
    [-obstructions list_of_layers]
```

#### Options

| Switch Name | Description |
| ----- | ----- |
| `-existing` | Flag to enable defining for existing routing solution. |
| `-name` | Defines a name to use when referring to this grid definition (defaults to `existing_grid`). |
| `-obstructions` | Specify the layers to add routing blockages, in order to avoid DRC violations. |

Example usage:

```tcl
define_pdn_grid -name main_grid -existing
```

### Define Power Switch Cell

> Define a power switch cell that will be inserted into a power grid 

Reviewer 2: ```suggestion
Define a power switch cell that will be inserted into a power grid.
```

| `-extend_to_core_ring` | Extend the stripes to the core PG ring. |
| `-pitch` | Value for the distance between each power/ground pair. |
| `-spacing` | Optional specification of the spacing between power/ground pairs within a single pitch (Default: pitch / 2). |
| `-offset` | Value for the offset of the stripe from the lower left corner of the design core area. |
| `-starts_with` | Specifies whether the first strap placed will be POWER or GROUND (Default: grid setting). |
| `-extend_to_boundary` | Extend the stripes to the boundary of the grid. |
| `-snap_to_grid` | Snap the stripes to the defined routing grid. |
| `-number_of_straps` | Number of power/ground pairs to add. |
| `-nets` | Limit straps to just this list of nets. |

Example usage:

```tcl
add_pdn_stripe -grid main_grid -layer metal1 -followpins
add_pdn_stripe -grid main_grid -layer metal2 -width 0.17 -followpins
add_pdn_stripe -grid main_grid -layer metal4 -width 0.48 -pitch 56.0 -offset 2 -starts_with GROUND
```

### Add Rings

> The `add_pdn_ring` command is used to define power/ground rings around a grid region. The ring structure is built using two layers that are orthogonal to each other. A power/ground pair will be added above and below the grid using the horizontal layer, with another power/ground pair to the left and right using the vertical layer. Together these 4 pairs of power/ground stripes form a ring around the specified grid. Power straps on these layers that are inside the enclosed region are extend to connect to the ring.

Reviewer 2: ```suggestion
The `add_pdn_ring` command defines power/ground rings around a grid region. The ring structure is built using two layers that are orthogonal to each other. A power/ground pair will be added above and below the grid using the horizontal layer, with another power/ground pair to the left and right using the vertical layer. These four pairs of power/ground stripes form a ring around the specified grid. Power straps on these layers that are inside the enclosed region are extended to connect to the ring.
```

| `-connect_to_pad_layers` | Restrict the pad pins layers to this list. |
| `-starts_with` | Specifies whether the first strap placed will be POWER or GROUND (Default: grid setting). |
| `-nets` | Limit straps to just this list of nets. |

Example usage: 

```tcl
add_pdn_ring -grid main_grid -layer {metal6 metal7} -widths 5.0 -spacings  3.0 -core_offset 5
```

### Add Connections

The `add_pdn_connect` command is used to define which layers in the power grid are to be connected together. During power grid generation, vias will be added for overlapping power nets and overlapping ground nets. The use of fixed vias from the technology file can be specified or else via stacks will be constructed using VIARULEs. If VIARULEs are not available in the technology, then fixed vias must be used.

```tcl
add_pdn_connect 
    -layers list_of_two_layers
    [-grid grid_name]
    [-cut_pitch pitch_value]
    [-fixed_vias list_of_fixed_vias]
    [-dont_use_vias list_of_vias]
    [-max_rows rows]
    [-max_columns columns]
    [-ongrid ongrid_layers]
    [-split_cuts split_cuts_mapping]
```

#### Options

| Switch Name | Description |
| ----- | ----- |

> | `-layers` | Layers to be connected where there are overlapping power or overlapping ground nets. |
> | `-grid` | Specifies the name of the grid definition to which this connection will be added (Default: Last grid created by `define_pdn_grid`). |
> | `-cut_pitch` | When the two layers are parallel e.g. overlapping stdcell rails, specify the distance between via cuts. |
> | `-fixed_vias` | List of fixed vias to be used to form the via stack. |
> | `-dont_use_vias` | List or pattern of vias to not use to form the via stack. |
> | `-max_rows` | Maximum number of rows when adding arrays of vias. |
> | `-max_columns` | Maximum number of columns when adding arrays of vias. |
> | `-ongrid` | List of intermediate layers in a via stack to snap onto a routing grid. |
> | `-split_cuts` | Specifies layers to use split cuts on with an associated pitch, for example `{metal3 0.380 metal5 0.500}`. |

Reviewer 2: ```suggestion
| `-layers` | Layers to be connected where there are overlapping power or overlapping ground nets. |
| `-grid` | Specifies the name of the grid definition to which this connection will be added (Default: Last grid created by `define_pdn_grid`). |
| `-cut_pitch` | When the two layers are parallel, e.g., overlapping stdcell rails, specify the distance between via cuts. |
| `-fixed_vias` | List of fixed vias to be used to form the via stack. |
| `-dont_use_vias` | List or pattern of vias to not use to form the via stack. |
| `-max_rows` | Maximum number of rows when adding arrays of vias. |
| `-max_columns` | Maximum number of columns when adding arrays of vias. |
| `-ongrid` | List of intermediate layers in a via stack to snap onto a routing grid. |
| `-split_cuts` | Specifies layers to use split cuts on with an associated pitch, for example `{metal3 0.380 metal5 0.500}`. |
```

add_pdn_connect -grid rotated_rams -layers {metal6 metal7}
```

### Repairing Power Grid vias after Detailed Routing

To remove vias which generate DRC violations after detailed placement and routing use `repair_pdn_vias`.

```tcl
repair_pdn_vias 
    [-all]
    [-net net_name]
```

#### Options

| Switch Name | Description |
| ----- | ----- |
| `-all` | Repair vias on all supply nets. |
| `-net` | Repair only vias on the specified net. |

### Useful Developer Commands

If you are a developer, you might find these useful. More details can be found in the [source file](../src/PdnGen.cc) or the [swig file](PdnGen.i).

| Command Name | Description |
| ----- | ----- |

> | `name_cmp` | Compare 2 input strings `obj1` and `obj2` if they are equal. |
> | `check_design_state` | Check if design is loaded. |
> | `get_layer` | Get the layer reference of layer name. |
> | `get_voltage_domains` | Gets a Tcl list of power domains in design. |
> | `match_orientation` | Checks if a given orientation `orient` is within a list of orientations `orients`. |
> | `get_insts` | Get Tcl list of instances. |
> | `get_masters` | Get Tcl list of masters. |
> | `get_one_to_two` | If a Tcl list has one element `{x}`, Tcl list `{x x}` is returned. If a Tcl list of two elements `{y y}`, list as is returned. Otherwise, for any other list lengths, error is triggered. |
> | `get_one_to_four` | Similar logic for above function, except the logic only works for lists of length one, two and four respectively. All other list lengths triggers error. |
> | `get_obstructions` | Get Tcl list of layers. |
> | `get_starts_with` | If value starts with `POWER`, return 1; else if value starts with `GROUND` return 0; else return error. |
> | `get_mterm` | Find master terminal. |
> | `get_orientations` | Get list of valid orientations. | 

Reviewer 2: ```suggestion
| `name_cmp` | Compare 2 input strings `obj1` and `obj2` if they are equal. |
| `check_design_state` | Check if the design is loaded. |
| `get_layer` | Get the layer reference of the layer name. |
| `get_voltage_domains` | Gets a Tcl list of power domains in design. |
| `match_orientation` | Checks if a given orientation `orient` is within a list of orientations `orients`. |
| `get_insts` | Get the Tcl list of instances. |
| `get_masters` | Get the Tcl list of masters. |
| `get_one_to_two` | If a Tcl list has one element `{x}`, Tcl list `{x x}` is returned. If a Tcl list of two elements `{y y}`, list as is returned. Otherwise, for any other list lengths, an error is triggered. |
| `get_one_to_four` | Similar logic for the above function, except the logic only works for lists of length one, two, and four, respectively. All other list lengths trigger errors. |
| `get_obstructions` | Get the Tcl list of layers. |
| `get_starts_with` | If value starts with `POWER`, return 1; else if value starts with `GROUND` return 0; else return error. |
| `get_mterm` | Find master terminal. |
| `get_orientations` | Get the list of valid orientations. | 
```

add_pdn_connect -grid rotated_rams -layers {metal6 metal7}
```

### Repairing Power Grid vias after Detailed Routing

To remove vias which generate DRC violations after detailed placement and routing use `repair_pdn_vias`.

```tcl
repair_pdn_vias 
    [-all]
    [-net net_name]
```

#### Options

| Switch Name | Description |
| ----- | ----- |
| `-all` | Repair vias on all supply nets. |
| `-net` | Repair only vias on the specified net. |

### Converting former PDNGEN configuration file to Tcl commands

To get an initial conversion from the former PDNGEN configuration file to the current tcl commands use `convert_pdn_config`.
This command will provide an initial set of commands based on the provided file, it is recommended that the user double-check
the conversion to ensure nothing was missed.

> ```tcl

Reviewer: this should not be documented as it's deprecated and will be removed soon and is not supported

pdngen
```

### Useful developer functions

If you are a developer, you might find these useful. More details can be found in the [source file](./src/PdnGen.cc) or the [swig file](./src/PdnGen.i).

```tcl
# source config file
source_config config

# get power nets
get_pdn_power_nets

# get ground nets
get_pdn_ground_nets

> ```

Reviewer: these are leftover from the old PDNgen, so they are not really developer functions.
Possibly:
```
gui::debug_renderer 1/0 < enable disable
gui::debug_renderer_update
```

Author: Noted, just saw your PR on removing these deprecated functions. 

Currently, these GUI commands are missing, but I think they better fit in the GUI tool section. What do you think? @gadfort

Reviewer: they are intended for developer use, so I think adding a GUI section to the developer section would be fine. General use users is discouraged.

## Regression tests

> There are a set of regression tests in `/test`. For more information, refer to this [section](../../README.md#regression-tests).

Reviewer 2: ```suggestion
There are a set of regression tests in `./test`. For more information, refer to this [section](../../README.md#regression-tests).
```
